src/topology/metric_space/dilation.lean

```diff
@@ -0,0 +1,274 @@
+/-
+Copyright (c) 2022 Hanting Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Dilations of emetric and metric spaces
+Authors: Hanting Zhang
+-/
+import topology.metric_space.antilipschitz
+import data.fun_like.basic
+
+/-!
+# Dilations
+
+We define dilations, i.e., maps between emetric spaces that
+satisfy `edist (f x) (f y) = r * edist x y`.
+
+Defines `ratio f`, which is the ratio of the dilation `f`
+
+Here `r : ℝ≥0`, so we do not exclude the degenerate case of dilations
+which collapse into constant maps. Since there is no `ℝ>0` API defined in mathlib,
+no matter where you choose to exclude r = 0, you will end up having to carry
+it around with you anyways. So statements that do need strict dilations should
+just say `f : dilation α β r` and `hr : r ≠ 0`.
+
+TODO: Introduce dilation equivs. Refactor the isometry API
+to match the `*_hom_class` API below.
+
+Since a lot of elementary properties don't require `eq_of_dist_eq_zero` we start setting up the
+theory for `pseudo_metric_space` and we specialize to `metric_space` when needed.
+-/
+
+noncomputable theory
+
+open function set
+open_locale topological_space ennreal nnreal
+
+section setup
+
+variables (α : Type*) (β : Type*) [pseudo_emetric_space α] [pseudo_emetric_space β]
+
+/-- A dilation is a map that uniformly scales the edistance between any two points.  -/
+structure dilation :=
+(to_fun : α → β)
+(edist_eq' : ∃ r : ℝ≥0, ∀ x y : α, edist (to_fun x) (to_fun y) = r * edist x y)
+
+attribute [nolint has_inhabited_instance] dilation
+
+/--
+`dilation_class F α β r` states that `F` is a type of `r`-dilations.
+
+You should extend this typeclass when you extend `dilation`.
+-/
+class dilation_class (F : Type*) (α β : out_param $ Type*)
+  [pseudo_emetric_space α] [pseudo_emetric_space β] extends fun_like F α (λ _, β) :=
+(edist_eq' : ∀ (f : F), ∃ r : ℝ≥0, ∀ (x y : α), edist (f x) (f y) = r * edist x y)
+
+attribute [nolint dangerous_instance] dilation_class.to_fun_like
+
+instance dilation.to_dilation_class :
+  dilation_class (dilation α β) α β :=
+{ coe := dilation.to_fun,
+  coe_injective' := λ f g h, by { cases f; cases g; congr', },
+  edist_eq' := λ f, dilation.edist_eq' f }
+
+instance : has_coe_to_fun (dilation α β) (λ _, α → β) := fun_like.has_coe_to_fun
+
+@[simp] lemma dilation.to_fun_eq_coe {f : dilation α β} : f.to_fun = (f : α → β) := rfl
+
+@[ext] theorem dilation.ext {f g : dilation α β} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+/-- Copy of a `dilation` with a new `to_fun` equal to the old one. Useful to fix definitional
+equalities. -/
+protected def dilation.copy (f : dilation α β) (f' : α → β) (h : f' = ⇑f) : dilation α β :=
+{ to_fun := f',
+  edist_eq' := h.symm ▸ f.edist_eq' }
+
+/-- The ratio of a dilation `f`. Uses `Exists.some` -/
+def ratio {α β} [pseudo_emetric_space α] [pseudo_emetric_space β] {F : Type*}
+  [dilation_class F α β] (f : F) : ℝ≥0 :=
+(dilation_class.edist_eq' f).some
```

Comment:
Any `def` defined in terms of `some` should be immediately followed by a corresponding lemma using `some_spec` to give the characteristic property of that `def`. In this case, there should probably also be a lemma or lemmas showing that `ratio` is equal to the distance ratio for any two points with nonzero finite distance (any two distinct points, in the metric space case) (and thus is the unique value satisfying the characteristic property).

Reply:
OK, so that first lemma is here as `edist_eq`; the uniqueness lemma is still needed.